### PR TITLE
[querier] Abort a query in limitedRoundTripper if parallelism is 0

### PR DESCRIPTION
**Pull request by cstyan**
Okay, I think I have this right :thinking: 

As far as I can tell, there isn't a defined limits default struct. However, we do parse the command line flags, and load those into a struct that is then passed as the defaults [here](https://github.com/grafana/loki/blob/1b5b8a31096784ac94981a0122b81e21ef6c5f78/cmd/loki/main.go#L36-L39). The default for `max_query_parallelism` is 32:
```
# Maximum number of queries that will be scheduled in parallel by the frontend.
# CLI flag: -querier.max-query-parallelism
[max_query_parallelism: <int> | default = 32]
```

We then modify that further with internally defined default Loki tenant sizes. 

So, it seems to me that there's not a possibility of the value of `max_query_parallelism` for a given Loki tenant to be `0` without that having been explicitly set in a config file.

Tl; DR, this PR modifies the `limitedRoundTripper`, which uses `max_query_paralllelism`, to abort a query early if a tenants query parallelism is less than 1. The current behaviour means that no additional Go routines would be scheduled to call the next step of the round trip chain **but** within the `limitedRoundTripper.RoundTrip` function we would not return until the context deadline (timeout) is hit within [this](https://github.com/grafana/loki/blob/1b5b8a31096784ac94981a0122b81e21ef6c5f78/pkg/querier/queryrange/limits.go#L302-L316) block.This is a less inefficient, but still not the ideal, solution to disable querying for a specific tenant ID.

Signed-off-by: Callum Styan <callumstyan@gmail.com>

**Checklist**
- [ ] Documentation added
- [x] Tests updated
- [ ] Add an entry in the `CHANGELOG.md` about the changes.

### CODE REVIEW

**Comment by cstyan** [comment]
I see the test failures, will fix in the morning :)

File: pkg/querier/queryrange/limits.go [hunk 1]
```diff
@@ -28,6 +28,10 @@ const (
 	limitErrTmpl = "maximum of series (%d) reached for a single query"
 )
 
+var (
+	ErrMaxQueryParalellism = fmt.Errorf("proto: max query parallelism")
```
Comment (cyriltovena) [inline pkg/querier/queryrange/limits.go:32]: Not sure about this error message. The user will receive it.

**Review by cyriltovena — APPROVED** [review]
LGTM.

I just want to make sure we understand this won't stop ALL queries.

Labels queries will still be available, and non filter too such as `{app="foo"}`. Although it's hard to abuse the system with those so I'm all good with this.

File: pkg/querier/queryrange/limits.go (hunk 1 above)
Comment (cstyan) [reply pkg/querier/queryrange/limits.go:32]: Any suggestions for alternatives?

**Comment by cstyan** [comment]
> I just want to make sure we understand this won't stop ALL queries.

Yep, this is just the first step in making disabling of queries for a tenant ID more efficient and explicit

> Labels queries will still be available, and non filter too such as {app="foo"}. Although it's hard to abuse the system with those so I'm all good with this.

Yep, looks to me like the `LimitedRoundTripper` is only used by log filter, series, metric, and instant metric queries. How come we don't use the parallelism limits for other query types?

File: pkg/querier/queryrange/limits.go (hunk 1 above)
Comment (cyriltovena) [reply pkg/querier/queryrange/limits.go:32]: Something saying querying is disable and you should get in touch with you cluster operator

**Comment by cyriltovena** [comment]
> How come we don't use the parallelism limits for other query types?

Because they are forwarded directly to querier, essentially parallelism 1 for those.
